### PR TITLE
Port VPP Fix to Megatron Hub

### PR DESCRIPTION
**Pull request by ZhiyuLi-Nvidia**
Fixes https://github.com/NVIDIA-NeMo/Megatron-Hub/issues/110
Follow the vpp refactoring in mcore, where change vp_stage (virtual pipeline stage/rank) from global state to local model/module attribute.

* pass vp stage to `MCoreGPTModel`
* use `is_pipeline_(first|last)_stage(ignore_virtual=False, vp_stage=i)` only in virtual pipeline parallelism
* Bump up to the same mcore commit as that in NeMo

### CODE REVIEW

File: src/megatron/hub/common/model_provider_mixin.py [hunk 1]
```diff
@@ -52,7 +52,7 @@ class ModelProviderMixin(abc.ABC, Generic[ModelT]):
     DEFAULT_CONFIG_FORMAT = "json"
 
     @abc.abstractmethod
-    def provide(self, pre_process=None, post_process=None) -> ModelT:
+    def provide(self, pre_process=None, post_process=None, vp_stage=None) -> ModelT:
```
Comment (ananthsub) [inline src/megatron/hub/common/model_provider_mixin.py:67]: this also needs to be kept in sync: https://github.com/NVIDIA-NeMo/Megatron-Hub/blob/7d195c05c78d4873dcb3475fdfb78c7e07045bdb/src/megatron/hub/core/models/model_provider.py#L46

**Comment by ZhiyuLi-Nvidia** [comment]
> @ZhiyuLi-Nvidia could you please rebase past https://github.com/NVIDIA-NeMo/Megatron-Hub/pull/108 ? fyi @marcromeyn for the model provider protocol needing to take in vp_stage now

@ananthsub I have rebased. And we have discussed offline it is fine to add take in vp_stage in model provider protocol.

Could you both take a review @ananthsub @marcromeyn ?

File: src/megatron/hub/common/model_provider_mixin.py (hunk 1 above)
Comment (ZhiyuLi-Nvidia) [reply src/megatron/hub/common/model_provider_mixin.py:67]: updated typing of `model_provider` and the following args description

```
 -   model_provider: Callable[[bool, bool], nn.Module],
 +   model_provider: Callable[[bool, bool, int], nn.Module],
...
-        model_provider: Callable that creates the model. Should accept
-            (pre_process, post_process) boolean arguments for pipeline parallelism
+        model_provider: Callable that creates the model. Should accept optional
+            pre_process(bool), post_process(bool), vp_stage(int) arguments for pipeline parallelism
```

File: src/megatron/hub/models/gpt_provider.py [hunk 2]
```diff
@@ -138,7 +138,7 @@ class GPTModelProvider(TransformerConfig, ModelProviderMixin[MCoreGPTModel]):
     bias_dropout_fusion: bool = field(default_factory=fusions.can_enable_bias_dropout_fusion)
     apply_rope_fusion: bool = field(default_factory=fusions.can_enable_apply_rope_fusion)
 
-    def provide(self, pre_process=None, post_process=None, tokenizer=None) -> MCoreGPTModel:
+    def provide(self, pre_process=None, post_process=None, tokenizer=None, vp_stage=None) -> MCoreGPTModel:
```
Comment (marcromeyn) [inline src/megatron/hub/models/gpt_provider.py]:
```suggestion
    def provide(self, pre_process=None, post_process=None, vp_stage=None, tokenizer=None) -> MCoreGPTModel:
```

Comment (marcromeyn) [reply src/megatron/hub/models/gpt_provider.py]: The tokenizer is specific to megatron-hub, so let's make it the last kwarg.

Comment (ZhiyuLi-Nvidia) [reply src/megatron/hub/models/gpt_provider.py]: Done.

File: tests/functional_tests/training/test_pretrain.py [hunk 3]
```diff
@@ -56,7 +56,8 @@ def test_pretrain_with_checkpoint(self, tmp_path):
 
             model_cfg = Llama32ModelProvider1B(
                 tensor_model_parallel_size=1,
-                pipeline_model_parallel_size=1,
+                pipeline_model_parallel_size=2,
+                virtual_pipeline_model_parallel_size=2,
                 context_parallel_size=1,
```
Comment (ananthsub) [inline tests/functional_tests/training/test_pretrain.py]: @ZhiyuLi-Nvidia instead of modifying this test, i'd recommend creating another test for vpp specifically. it can be with a smaller model for testing: https://github.com/NVIDIA-NeMo/Megatron-Hub/blob/06613212d922b523bd352bcee03b91dd592c4cba/src/megatron/hub/models/gpt_provider.py#L289-L302

Comment (ZhiyuLi-Nvidia) [reply tests/functional_tests/training/test_pretrain.py]: Sure.

File: src/megatron/hub/models/gpt_provider.py (hunk 2 above)
Comment (ZhiyuLi-Nvidia) [reply src/megatron/hub/models/gpt_provider.py]: Done.

File: tests/functional_tests/training/test_pretrain.py (hunk 3 above)
Comment (ZhiyuLi-Nvidia) [reply tests/functional_tests/training/test_pretrain.py]: Added a vpp test of `Llama32ModelProvider1B` model. Chose this model since we have the training loss from the same model in pretrain test to benchmark against.

Verified the correctness with a close losses with and without virtual pipeline.
